### PR TITLE
Adopt ieee-idams gem for parsing ieee-rawbib data

### LINKED ISSUE #42: Adopt ieee-idams gem for parsing ieee-rawbib data
https://github.com/relaton/ieee-idams is now completed and fully round-trips (all) ieee-rawbib data!

This is ready to go.

### PR DESCRIPTION
**Pull request by andrew2net**
Fixes #42

### CODE REVIEW

File: lib/relaton_ieee/idams_parser.rb [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+      t << RelatonBib::TypedTitleString.new(content: content, type: "main")
+    end
+
+    #
+    # Parse date
+    #
+    # @return [Array<RelatonBib::BibliographicDate>]
+    #
+    def parse_date
+      dates = @doc.volume.article.first.articleinfo.date.map do |date|
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:75]: Why not move the date parsing to the ieee-idams gem?

File: lib/relaton_ieee/idams_parser.rb [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    end
+
+    #
+    # Convert date string with month name to numeric date
+    #
+    # @param [String] date source date
+    #
+    # @return [String] numeric date
+    #
+    def parse_date_string(date)
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:101]: Why not move the date parsing to the ieee-idams gem?

File: lib/relaton_ieee/idams_parser.rb [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    #
+    # Parse contributors
+    #
+    # @return [Array<RelatonBib::ContributionInfo>]
+    #
+    def parse_contributor
+      contrib = @doc.publicationinfo.publisher
+      name = contrib.publishername
+      addr = create_addres contrib.address
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:135]: Typo in `creat_addres`

File: lib/relaton_ieee/idams_parser.rb [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    end
+
+    #
+    # Parse copyright
+    #
+    # @return [Array<RelatonBib::CopyrightAssociation>]
+    #
+    def parse_copyright
+      @doc.publicationinfo.copyrightgroup.copyright.map do |c|
+        owner = c.holder.split("/").map do |own|
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:193]: Move this to the ieee-idams gem?

File: lib/relaton_ieee/idams_parser.rb [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    end
+
+    #
+    # Parse status
+    #
+    # @return [RelatonIee::DocumentStatus, nil]
+    #
+    def parse_docstatus
+      st = parse_standard_modified
+      return unless %w[Draft Approved Superseded Withdrawn].include?(st)
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:207]: This should be in the ieee-idams gem.

File: lib/relaton_ieee/idams_parser.rb [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      url = "https://ieeexplore.ieee.org/document/#{id}"
+      [RelatonBib::TypedUri.new(content: url, type: "src")]
+    end
+
+    #
+    # Parse keyword
+    #
+    # @return [Array<Strign>]
+    #
+    def parse_keyword
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:246]: This should be in the ieee-idams gem?

File: lib/relaton_ieee/idams_parser.rb [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+      @doc.volume.article[0].articleinfo.keywordset[0]&.keyword&.map &:keywordterm
+    end
+
+    #
+    # Parse ICS
+    #
+    # @return [Array<RelatonBib::ICS>]
+    #
+    def parse_ics
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:255]: This should be in the ieee-idams gem?

File: lib/relaton_ieee/idams_parser.rb [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    def parse_holdstatus
+      @doc.publicationinfo.holdstatus
+    end
+
+    #
+    # Parse doctype
+    #
+    # @return [String] doctype
+    #
+    def parse_doctype
```
Comment (ronaldtse) [inline lib/relaton_ieee/idams_parser.rb:314]: This should be in the ieee-idams gem?

File: lib/relaton_ieee/data_fetcher.rb [hunk 9]
```diff
@@ -88,27 +88,25 @@ def read_zip(file)
...
         return
       end
-      amsid = doc.at("./publicationinfo/amsid").text
+      amsid = doc.publicationinfo.amsid
       if backrefs.value?(bib.docidentifier[0].id) && /updates\.\d+/ !~ filename
         oamsid = backrefs.key bib.docidentifier[0].id
         Util.warn "Document exists ID: `#{bib.docidentifier[0].id}` AMSID: " \
              "`#{amsid}` source: `#{filename}`. Other AMSID: `#{oamsid}`"
-        if bib.docidentifier[0].id.include?(doc.at("./publicationinfo/stdnumber").text)
+        if bib.docidentifier[0].id.include?(doc.publicationinfo.stdnumber)
```
Comment (ronaldtse) [inline lib/relaton_ieee/data_fetcher.rb:109]: This should be in the ieee-idams gem?

**Review by ronaldtse — CHANGES_REQUESTED** [review]

**Comment by andrew2net** [comment]
@ronaldtse do you mean we should move the whole parser to ieee-idams so it returns `IeeeBibliographicItem`?

**Comment by andrew2net** [comment]
@ronaldtse or we should move the data fetching feature to the ieee-idams? So we will use ieee-idams instead relaton-iee in relaton-data-ieee GHA, right?

**Comment by ronaldtse** [comment]
That's not what I mean... I want the IDAMS data to be extracted cleanly so that relaton-ieee does not do low level data formatting and string parsing. So those low level features should be implemented in ieee-idams, and relaton-ieee only builds the Relaton model using direct methods in ieee-idams (as much as possible).

**Comment by andrew2net** [comment]
We have IEEE data model inside relaton-ieee gem, and the data is converted to the model inside the gem. If we move converting functionality to ieee-idams then:
- ieee-idams will depend on relaton-ieee and relaton-ieee will depend on ieee-idams which is not good
- the 2 gems become coupled, i.e. if we change model in relaton-ieee we may need to update the converting functionality in ieee-idams

I think if we need to move covering functionality to ieee-idams we should move it with model.

**Comment by andrew2net** [comment]
@ronaldtse ok, I'll try to move to ieee-idams all the low level data transformations that don't involve creating Relaton-model objects.

**Comment by ronaldtse** [comment]
> ieee-idams will depend on relaton-ieee

indeed we do not want that. The ieee-idams gem must only deal with that data’s concern, not Relaton items.

File: lib/relaton_ieee/data_fetcher.rb (hunk 9 above)
Comment (andrew2net) [reply lib/relaton_ieee/data_fetcher.rb:109]: I don't think so. The `backrefs` variable is a list of relation references which collected through all the documents and used later to create relations. The ieee-idams operates on a single document and shouldn't be involved in collection functionality.

Comment (andrew2net) [reply lib/relaton_ieee/data_fetcher.rb]: The update matches against the source filename path. It's not any kind of file content.

**Review by ronaldtse — APPROVED** [review]
@andrew2net thank you! Please feel free to merge and release when ready.

**Pull request merged**